### PR TITLE
Trie lib (part. 2)

### PR DESCRIPTION
**Pull request by ClementWalter**

### CODE REVIEW

File: cairo/ethereum/cancun/trie.cairo [hunk 1]
```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
...
+    local value: Bytes;
+    local value_set: felt;
+
+    tempvar branch = branch;
+    tempvar dict_ptr = obj.value.dict_ptr_start;
+
+    loop:
+    let branch = cast([ap - 2], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 1], BytesBytesDictAccess*);
+    // The verifier just needs to make sure that whatever case we are in is properly asserted.
```
Comment (enitrat) [inline cairo/ethereum/cancun/trie.cairo:504]:
```suggestion
    // The assertions on the values returned by this nondet memory assignment are verified in the corresponding branches
```

File: cairo/ethereum/cancun/trie.cairo [hunk 2]
```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
...
+    let value_set = [fp + 3];
+
+    // Fill value_set if it's not set yet. This is just to be able to test against 1
+    // as this would raise if the memory is empty.
+    %{ ids.value_set = memory.get(fp + 3) or 0 %}
+    if (value_set != 1) {
+        let (data: felt*) = alloc();
+        tempvar empty_bytes = Bytes(new BytesStruct(data, 0));
+        assert value = empty_bytes;
+    }
```
Comment (enitrat) [inline cairo/ethereum/cancun/trie.cairo:583]: possible inconsistency here:

value_case => value_set == 1, value = dict_ptr.new_value
but the prover sets: value_set = 0 here

idea: if value_set != 1, `assert cast(value, felt) != 0` ?

Comment (enitrat) [reply cairo/ethereum/cancun/trie.cairo:583]: ah, no ok: there's a speific assigment of `empty_bytes` if the value is not set.
can't we not just assert that the prover did not return the wrong response in the hint?

File: cairo/ethereum/cancun/trie.cairo [hunk 3]
```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
...
+    if (value_15.value.len != 0) {
+        assert value = value_15;
+        assert value_set = 1;
+    }
+    %{ ids.value_set = memory.get(fp + 2) or 0 %}
+    if (value_set != 1) {
+        let (data: felt*) = alloc();
+        tempvar empty_bytes = Bytes(new BytesStruct(data, 0));
+        assert value = empty_bytes;
+    }
```
Comment (enitrat) [inline cairo/ethereum/cancun/trie.cairo:698]: (same as previous comment regarding value_set being wrongly returned by the prover and checks on existing memory rather than creating a new segment)

File: cairo/ethereum/rlp.cairo [hunk 4]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
-        assert [dst] = 0xC0 + len;
-        tempvar result = Bytes(new BytesStruct(dst, 1 + len));
-        return result;
-    }
-
-    let (len_joined_encodings: felt*) = alloc();
-    let len_joined_encodings_len = felt_to_bytes_little(len_joined_encodings, len);
+    assert [dst] = 0x80 + 32;
+    uint256_to_bytes32_little(dst + 1, [code_hash.value]);
+    let dst = dst + 33;
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:308]:
```suggestion
    let code_hash_len = 32;
    assert [dst] = 0x80 + code_hash_len;
    uint256_to_bytes32_little(dst + 1, [code_hash.value]);
    let encoded_code_hash_len = 33;
    let dst = dst + encoded_code_hash_len;
```

File: cairo/ethereum/rlp.cairo [hunk 5]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
-    let dst = dst - 1 - len_joined_encodings_len;
-    assert [dst] = 0xF7 + len_joined_encodings_len;
-    // Copy the length encoding
-    memcpy(dst + 1, len_joined_encodings, len_joined_encodings_len);
+    let len = dst - dst_start - PREFIX_LEN_MAX;
+    let dst = dst_start + PREFIX_LEN_MAX;
+    let offset = _encode_prefix_len(dst, len);
 
-    tempvar result = Bytes(new BytesStruct(dst, 1 + len_joined_encodings_len + len));
+    tempvar result = Bytes(new BytesStruct(dst - offset, offset + len));
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:314]: hard to follow what's going on here, suggesting something like:

```suggestion
    let body_len = dst - dst_start - PREFIX_LEN_MAX;
    let body_ptr = dst_start + PREFIX_LEN_MAX;
    let prefix_len = _encode_prefix_len(body_ptr, body_len);

    tempvar result = Bytes(new BytesStruct(body_ptr - prefix_len, prefix_len + body_len));
```

File: cairo/ethereum/rlp.cairo [hunk 6]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
 
-    // Write the length encoding
-    // Length encoding is 1 byte for the prefix and then the length in little endian
-    let dst = dst - len_joined_encodings_len;
-    assert [dst] = 0xF7 + len_joined_encodings_len;
-    // Copy the length encoding
-    memcpy(dst + 1, len_joined_encodings, len_joined_encodings_len);
-
-    tempvar result = Bytes(new BytesStruct(dst, 1 + len_joined_encodings_len + len));
+    tempvar result = Bytes(new BytesStruct(dst - offset, offset + len));
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:346]: same suggestion as above
```suggestion
    let body_len = dst - dst_start - PREFIX_LEN_MAX;  
    let body_ptr = dst_start + PREFIX_LEN_MAX;  
    let prefix_len = _encode_prefix_len(body_ptr, body_len);  

    tempvar result = Bytes(new BytesStruct(body_ptr - prefix_len, prefix_len + body_len));  
```

File: cairo/ethereum/rlp.cairo [hunk 7]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
+    let dst = dst_start + PREFIX_LEN_MAX;
+    let offset = _encode_prefix_len(dst, len);
 
-    let cond = is_le(len, 0x38 - 1);
-    if (cond != 0) {
-        assert [dst] = 0xC0 + len;
-        tempvar result = Bytes(new BytesStruct(dst, 1 + len));
-        return result;
-    }
+    tempvar result = Bytes(new BytesStruct(dst - offset, offset + len));
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:366]:
```suggestion
    let body_len = dst - dst_start - PREFIX_LEN_MAX;  
    let body_ptr = dst_start + PREFIX_LEN_MAX;  
    let prefix_len = _encode_prefix_len(body_ptr, body_len);  

    tempvar result = Bytes(new BytesStruct(body_ptr - prefix_len, prefix_len + body_len));  
```

File: cairo/ethereum/rlp.cairo [hunk 8]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
+    let bloom_len = _encode_bloom(dst, raw_receipt.value.bloom);
+    let dst = dst + bloom_len;
+    let logs_len = _encode_tuple_log(dst, raw_receipt.value.logs);
+    let dst = dst + logs_len;
+
+    let len = dst - dst_start - PREFIX_LEN_MAX;
+    let dst = dst_start + PREFIX_LEN_MAX;
+    let offset = _encode_prefix_len(dst, len);
+
+    tempvar result = Bytes(new BytesStruct(dst - offset, offset + len));
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:404]:
```suggestion
    let body_len = dst - dst_start - PREFIX_LEN_MAX;  
    let body_ptr = dst_start + PREFIX_LEN_MAX;  
    let prefix_len = _encode_prefix_len(body_ptr, body_len);  

    tempvar result = Bytes(new BytesStruct(body_ptr - prefix_len, prefix_len + body_len));  
```

File: cairo/ethereum/rlp.cairo [hunk 9]
```diff
@@ -284,127 +290,143 @@ func encode_account{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: K
...
+    let address_len = _encode_address(dst, raw_withdrawal.value.address);
+    let dst = dst + address_len;
+    let amount_len = _encode_u256(dst, raw_withdrawal.value.amount);
+    let dst = dst + amount_len;
+
+    let len = dst - dst_start - PREFIX_LEN_MAX;
+    let dst = dst_start + PREFIX_LEN_MAX;
+    let offset = _encode_prefix_len(dst, len);
+
+    tempvar result = Bytes(new BytesStruct(dst - offset, offset + len));
```
Comment (enitrat) [inline cairo/ethereum/rlp.cairo:426]:
```suggestion
    let body_len = dst - dst_start - PREFIX_LEN_MAX;  
    let body_ptr = dst_start + PREFIX_LEN_MAX;  
    let prefix_len = _encode_prefix_len(body_ptr, body_len);  

    tempvar result = Bytes(new BytesStruct(body_ptr - prefix_len, prefix_len + body_len));  
```

**Review by enitrat — COMMENTED** [review]
💯

File: cairo/ethereum/cancun/trie.cairo [hunk 10]
```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
...
+
+    tempvar branch = branch;
+    tempvar dict_ptr = obj.value.dict_ptr_start;
+
+    loop:
+    let branch = cast([ap - 2], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 1], BytesBytesDictAccess*);
+    // The verifier just needs to make sure that whatever case we are in is properly asserted.
+    tempvar is_nibble_case = nondet %{ memory.get(ids.dict_ptr.key.value.data + ids.level) == ids.nibble %};
+    tempvar is_value_case = nondet %{ int(ids.dict_ptr.key.value.len == ids.level) %};
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply cairo/ethereum/cancun/trie.cairo:558]: but what does this change? because each branch checks its own condition, so if the prover wants to go in `value_case` while it's not possible, it won't be able to satisfy the constraint `assert dict_ptr.key.value.len = level;`

[the remark this answers is not in the document]
Comment (ClementWalter) [reply cairo/ethereum/cancun/trie.cairo:558]: note that `is_zero(a-b)` would have done it as well. Overall I could remove these two hints, handle the value_case first, and otherwise do the check nibble or not

File: cairo/ethereum/cancun/trie.cairo (hunk 1 above)
Comment (ClementWalter) [reply cairo/ethereum/cancun/trie.cairo:556]: I think that it's important to strength the fact that the verifier doesn't know nor need to verify at each iteration which branch is the right one, but only that whatever branch the prover chose is sufficiently constraint

File: cairo/ethereum/cancun/trie.cairo (hunk 2 above)
Comment (ClementWalter) [reply cairo/ethereum/cancun/trie.cairo:635]: > but the prover sets: value_set = 0 here

this is not possible, the prover cannot override a value. if value_set as been set to 1, then it's 1.

Comment (ClementWalter) [reply cairo/ethereum/cancun/trie.cairo:635]: I don't understand what you mean by 

> can't we not just assert that the prover did not return the wrong response in the hint?

the idea is that the value is either constraint (value case) or not. I just want to check if it's not constraint, and if not, then put an empty bytes.

It's also possible to do it with no hints and a pointer. I could do

```
local value_set: felt*

// ...

value_case:
assert value_set = alloc();

// ...

if (cast(value_case, felt) == 0) {
    // set empty bytes
}
```

but it's a bit hacky as it uses the fact that you can use a relocatable to have set or not, while we want a bool

File: cairo/ethereum/cancun/trie.cairo (hunk 10 above)
[the remark this answers is not in the document]
Comment (enitrat) [reply cairo/ethereum/cancun/trie.cairo:558]: I think it's best removing the hints and just do a is_zero
When it comes to values provided through hints, I believe the habit of having explicit and clear assertions outweighs  the risk of "over-optimizing" by relying on other implied assumptions.

In that case I prefer seeing:
```
res_0 = %{something happening here%}
res_1 = %{..%}
assert res_0 * res_1  = 1
```
which clearly outlines the valid states: either both are 0, or one of them is 0, but having both of them == 1 is an invalid state

**Review by enitrat — APPROVED** [review]
